resource/config_map.go

```diff
@@ -0,0 +1,198 @@
+	}
+
+	var current *corev1.ConfigMap
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding current ConfigMap %#q", printName))
+
+		current, err = c.k8sClient.CoreV1().ConfigMaps(c.namespace).Get(c.name, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find current ConfigMap %#q", printName))
+
+		} else if err != nil {
+			return microerror.Mask(err)
+
+		} else {
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found current ConfigMap %#q", printName))
+		}
+	}
+
+	if current != nil {
+		c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("updating ConfigMap %#q", printName))
+
+		toUpdate := newConfigMapToUpdate(current, desired)
+
+		if toUpdate != nil {
+			_, err := c.k8sClient.CoreV1().ConfigMaps(c.namespace).Update(toUpdate)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("updated ConfigMap %#q", printName))
+
+		} else {
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("ConfigMap %#q is up to date", printName))
+		}
+
+	} else {
+		c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("creating ConfigMap %#q", printName))
+
+		_, err := c.k8sClient.CoreV1().ConfigMaps(c.namespace).Create(desired)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("created ConfigMap %#q", printName))
+	}
+
+	return nil
+}
+
+func (c *ConfigMap) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	var err error
+
+	printName := newPrintName(c.namespace, c.name)
+
+	{
+		c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("deleting ConfigMap %#q", printName))
+
+		err = c.k8sClient.CoreV1().ConfigMaps(c.namespace).Delete(c.name, &metav1.DeleteOptions{})
+		if apierrors.IsNotFound(err) {
+			c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("ConfigMap %#q already deleted", printName))
+
+		} else if err != nil {
+			return microerror.Mask(err)
+
+		} else {
+			c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("deleted ConfigMap %#q", printName))
+		}
+	}
+
+	return nil
+}
+
+// newConfigMapToUpdate creates a new ConfigMap to update based on current and
+// desired one. If there is no update nil is returned.
+func newConfigMapToUpdate(current, desired *corev1.ConfigMap) *corev1.ConfigMap {
+	merged := current.DeepCopy()
+
+	if reflect.DeepEqual(current.Annotations, desired.Annotations) {
+		merged.Labels = desired.Annotations
+	}
+	if reflect.DeepEqual(current.Labels, desired.Labels) {
+		merged.Labels = desired.Labels
+	}
+
+	if reflect.DeepEqual(current.BinaryData, desired.BinaryData) {
+		merged.BinaryData = desired.BinaryData
+	}
+	if reflect.DeepEqual(current.Data, desired.Data) {
+		merged.Data = desired.Data
+	}
+
+	if reflect.DeepEqual(current, merged) {
+		return nil
+	}
+
+	return merged
+}
```

Comment:
There should be also e2e test for that. At least some basic scenario. This is cool that we can write single isolated test for that and use the resource in other operators.